# Grade Entry System

**Author**: Brady Woods  
**Course**: BGEN632 Grad. Intro. to Python  
**Term**: Spring 2025  
**Date**: April 7, 2025  

This notebook produces a user menu that stores, queries, and adds student information. It will allow the user to view existing student information, query out GPA details, add in new students along with their information, and add in student violations. It is a one stop shop for learning management.

## Initialize student records


In [1]:
students = [                                   #Assigning students and their information
    {"first_name": "Carmen", "last_name": "Berzatto", "major": "Marketing", "gpa": 3.91, "violation": None},
    {"first_name": "Sydney", "last_name": "Adamu", "major": "Business", "gpa": 3.72, "violation": None},
    {"first_name": "Neil", "last_name": "Fak", "major": "Communications", "gpa": 2.95, "violation": None},
    {"first_name": "Marcus", "last_name": "Brooks", "major": "Economics", "gpa": 3.23, "violation": None},
]


violations_list = [                              #Assigning potential violations
    "Distributing Course Materials",
    "Excessive Plagiarism",
    "Excessive Absences",
    "Using Technology During Exam",
    "Collabterating on Exam"
]


## Menu loop


In [2]:
while True:                                     #While loop to keep the program running
    print("\n Student Management System")
    print("1. Student Information Query")
    print("2. Student Information Entry")
    print("3. Query GPA Metrics")
    print("4. Academic Violations")
    print("5. Quit Program")
    choice = input("\nSelect an option (1-5): ")

    if choice == '1':                           #Choice for querying student information
        print("\n All Student Information")
        for student in students:
            print(f"{student['first_name']} {student['last_name']} | Major: {student['major']} | GPA: {student['gpa']} | Violation: {student['violation']}")

    elif choice == '2':                         #Choice for entering new student information            
        print("\n Add New Student")
        first = input("First name: ")
        last = input("Last name: ")
        major = input("Major: ")
        gpa = float(input("GPA: "))
        students.append({"first_name": first, "last_name": last, "major": major, "gpa": gpa, "violation": None})
        print("Student added successfully!")

    elif choice == '3':                         #Choice for querying GPA metrics
        gpas = [s['gpa'] for s in students]
        print("\n--- GPA Metrics ---")
        print(f"Max GPA: {max(gpas)}")
        print(f"Min GPA: {min(gpas)}")
        print(f"Average GPA: {sum(gpas)/len(gpas):.2f}")

    elif choice == '4':                        #Choice for assigning academic violations
        print("\n Assign Academic Violations")
        for i, student in enumerate(students):
            print(f"{i+1}. {student['first_name']} {student['last_name']}")
        selected = int(input("Select a student by number: ")) - 1
        print("Select a violation:")
        for i, v in enumerate(violations_list):
            print(f"{i+1}. {v}")
        v_choice = int(input("Violation number: ")) - 1
        students[selected]['violation'] = violations_list[v_choice]
        print("Violation assigned successfully.")

    elif choice == '5':                       #Choice for quitting the program
        print("\nExiting program. Goodbye!")
        break

    else:
        print("\nInvalid selection. Try again.")



 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

--- GPA Metrics ---
Max GPA: 3.91
Min GPA: 2.95
Average GPA: 3.45

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

 Add New Student
Student added successfully!

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

 Add New Student
Student added successfully!

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

--- GPA Metrics ---
Max GPA: 4.0
Min GPA: 2.72
Average GPA: 3.42

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

 Assign Academic Violations
1. Carmen Berzat

## I have copied and pasted the full output into a markdown cell below. 


 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

--- GPA Metrics ---
Max GPA: 3.91
Min GPA: 2.95
Average GPA: 3.45

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

 Add New Student
Student added successfully!

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

 Add New Student
Student added successfully!

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

--- GPA Metrics ---
Max GPA: 4.0
Min GPA: 2.72
Average GPA: 3.42

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

 Assign Academic Violations
1. Carmen Berzatto
2. Sydney Adamu
3. Neil Fak
4. Marcus Brooks
5. Brady Woods
6. Jonathan Woods
Select a violation:
1. Distributing Course Materials
2. Excessive Plagiarism
3. Excessive Absences
4. Using Technology During Exam
5. Collabterating on Exam
Violation assigned successfully.

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

--- GPA Metrics ---
Max GPA: 4.0
Min GPA: 2.72
Average GPA: 3.42

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

 Assign Academic Violations
1. Carmen Berzatto
2. Sydney Adamu
3. Neil Fak
4. Marcus Brooks
5. Brady Woods
6. Jonathan Woods
Select a violation:
1. Distributing Course Materials
2. Excessive Plagiarism
3. Excessive Absences
4. Using Technology During Exam
5. Collabterating on Exam
Violation assigned successfully.

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

 All Student Information
Carmen Berzatto | Major: Marketing | GPA: 3.91 | Violation: None
Sydney Adamu | Major: Business | GPA: 3.72 | Violation: None
Neil Fak | Major: Communications | GPA: 2.95 | Violation: None
Marcus Brooks | Major: Economics | GPA: 3.23 | Violation: None
Brady Woods | Major: Economics | GPA: 4.0 | Violation: Excessive Absences
Jonathan Woods | Major: Mathematics | GPA: 2.72 | Violation: Using Technology During Exam

 Student Management System
1. Student Information Query
2. Student Information Entry
3. Query GPA Metrics
4. Academic Violations
5. Quit Program

Exiting program. Goodbye!


## Comments
One thing I struggled with a little in this assignment was in the adding violations portion. While testing out my code, I tried adding multiple violations to one student and it would override the previous violation and only display the most recent entry. My thought to fix this would be to create a new list for second and third violations for example, but I was curious if there would be a more efficient way to store this information; for example maybe a list inside of a list in this case? I tried not to get hung up on this as it wasn't a requirement for the assignment but from a data perspective it seens like a silly reason to lose data.